### PR TITLE
Lecture 26: Gaussian processes (GPs) and elements of meta-learning

### PR DESCRIPTION
**Pull request by anjalief**
https://anjalief.github.io/pgm-spring-2019/notes/lecture/

### CODE REVIEW

File: _posts/2019-04-26-lecture.md [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+
+$$ \log p(\mathbf{y} | \mathbb{x}, \mathbf{w}) \propto -\frac{1}{2 \sigma^2} \sum_{i=1}^n(f(\mathbb{x_i}, \mathbf{w}) - y(\mathbb{x_i})^2) - \lambda \mathbb{w}^T\mathbb{w}  $$
+
+However, this introduces new questions: how do we define complexity? and how much should we penalize complexity? In practice, we control the penalty by setting  $\lambda$ using cross-validation
+
+**Bayesian Approach**
+
+We can describe our data and models using Bayes' Rule:
+
+$$posterior = \frac{likelihood * prior}{marginal\ likelihood}$$
```
Comment (pliang279) [inline _posts/2019-04-26-lecture.md]: do \textrm for normal text in math mode

File: _posts/2019-04-26-lecture.md [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+\begin{aligned}
+\left[ f(x_1) ..f(x_N) \right] \sim \mathcal{N}(0, K)
+\end{aligned}
+</d-math>
+
+where $K$ is defined by,
+
+<d-math block>
+\begin{aligned}
+K_{ij} = cov(f(x_i), f(x_j)) = k(x_i, x_j) = 1 + x_bx_c
```
Comment (pliang279) [inline _posts/2019-04-26-lecture.md]: do \textrm for normal text in math mode

File: _posts/2019-04-26-lecture.md [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+  </figcaption>
+</figure>
+
+**Kernel Example: RBF**
+
+RBF (Radial Basis Function) is the most popular kernel used with Gaussian processes. It is given as
+<d-math block>
+\begin{aligned}
+k_{\text{RBF}(x, x')} &= \text{cov} (f(x), f(x')) \\
+&= a^2 \exp(-\frac{\Vert x-x'\Vert^2}{2\ell^2})
```
Comment (pliang279) [inline _posts/2019-04-26-lecture.md]: \left( and \right)

File: _posts/2019-04-26-lecture.md [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+</d-math>
+
+ This parameterization makes it easier to apply gaussian processes on a wide range of tasks, for example, sequential data. A sequence can be encoded using a recurrent neural network and a kernel function can be applied to the encoded representation. For more details, please refer to (cite) which uses a Gaussian process on top of a LSTM to predict the prediction of lead by vehicles. 
+
+ **The Scalability Issue**
+
+ GP inference requires computing inverse and determinants of huge covariance matrices over the entire training data which can be computationally intenstive.
+
+ * Inference requires a $(k_\theta + \sigma^2 \mathbf{I})^{-1} \mathbf{y}$ step and,
+ * Learning requires a $\log | k_\theta + \sigma^2 \mathbf{I}|$ step,
```
Comment (pliang279) [inline _posts/2019-04-26-lecture.md]: this does not render correctly

File: _posts/2019-04-26-lecture.md [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+* There is a variety of approximation methods to GPs that can bring down the learning and inference cost to $O(n)$ and $O(1)$, respectively
+* Many new libraries based on Tensorflow, PyTorch, Keras -- despite computational constraints, GPs are certainly quite popoular
+
+## Meta-learning and Neural Processes
+
+So far, we assume that data was generated by a single function. What if there are multiple data-generating functions, and each time we get only a few points from one of them. Can we identify it?
+
+**Definition of meta-learning**
+
+In standard learning, given a distribution over examples (single task), we learn a function that minimizes the loss $\hat \phi = argmin_{\phi} E_{z \sim D}[l(f_{\phi}(z))]$
```
Comment (pliang279) [inline _posts/2019-04-26-lecture.md]: argmin in \textrm

File: _posts/2019-04-26-lecture.md [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+In learning-to-learn, given a distribution over tasks, output an adaptation rule that can be used at test time to generalize from a task description: $\hat \theta = argmin_{\theta} E_{T \sim P}[L_T(g_{\theta}(T))]$ where $L_T(g_{\theta}(T)) \coloneqq E_{z \sim D}[l(f_{\phi}(z))]$, $\phi \coloneqq g_{\theta}(T)$
+
+Examples:
+
+* Few-shot image classification: multiple datasets, each of which includes a few examples of each class.
+* Few-shot user-specific recommendation: recommend twitter posts to users; each user has different set of likes and dislikes (binary prediction)
+* Contextual interpretability: interpretable linear model conditioned on images
+* One-shot imitation learning: imitate a single behaviour from demonstrations; give a new demonstration, produce a policy to quickly imitate the demonstration
+
+**Conditional Neural Processes**
```
Comment (pliang279) [inline _posts/2019-04-26-lecture.md:392]: add references and expand more, a figure would help

File: _posts/2019-04-26-lecture.md [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+Try to produce representations for observable inputs and labels, $r$. The representations are aggregated and fed to function $g$ for prediction. So we can produce different function $g$'s given different sets of training data. This is similar to Gaussian Processes.
+
+**Attentive Neural Processes**
```
Comment (pliang279) [inline _posts/2019-04-26-lecture.md:395]: add references and expand more, a figure would help

**Review by pliang279 — CHANGES_REQUESTED** [review]
hi everyone, please make these small changes, thanks!

**Comment by anjalief** [comment]
I've made the requested changes to the first third in https://github.com/sailinglab/pgm-spring-2019/pull/34/commits/2e241c5bc31c830112d2418148dd0465add96260

**Comment by alshedivat** [comment]
@anjalief, @pliang279, can you iterate on this quickly and merge by EOD today? Thanks.

**Comment by anjalief** [comment]
@pliang279 I've made additional corrections to the lines that didn't render properly in https://github.com/anjalief/pgm-spring-2019/commit/43674d8c0ce06f2e6f9f0da05fb8be1d0fabc6fa

**Pull request merged**